### ISSUE TITLE
DNS issues in linkerd2 > 2.8.1

### ISSUE THREAD

**Issue by hwaastad** [linkerd/linkerd2#5932]
### What is the issue?
linkerd pods does not start due to dns issues.
Linkerd2 2.8.1 is working fine

### Logs, error output, etc
controller-proxy
```text
 0.004341s]  INFO ThreadId(01) linkerd2_proxy: Local identity is linkerd-controller.linkerd.serviceaccount.identity.linkerd.cluster.local
[     0.004348s]  INFO ThreadId(01) linkerd2_proxy: Identity verified via linkerd-identity-headless.linkerd.svc.cluster.local:8080 (linkerd-identity.linkerd.serviceaccount.identity.linkerd.cluster.local)
[     0.004354s]  INFO ThreadId(01) linkerd2_proxy: Destinations resolved via linkerd-dst-headless.linkerd.svc.cluster.local:8086 (linkerd-destination.linkerd.serviceaccount.identity.linkerd.cluster.local)
[     0.012927s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.cluster.local. type: SRV class: IN
[     0.016730s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.cluster.local. type: SRV class: IN
[     2.768909s]  INFO ThreadId(02) daemon:admin{listen.addr=0.0.0.0:4191}:accept{client.addr=127.0.0.1:56226 target.addr=127.0.0.1:4191}: linkerd_app_core::serve: Connection closed error=TLS detection timeout
[     3.770999s]  INFO ThreadId(02) daemon:admin{listen.addr=0.0.0.0:4191}:accept{client.addr=127.0.0.1:56238 target.addr=127.0.0.1:4191}: linkerd_app_core::serve: Connection closed error=TLS detection timeout
[     5.015220s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.cluster.local. type: SRV class: IN
[     5.027275s]  WARN ThreadId(01) inbound:accept{client.addr=10.42.5.36:37204 target.addr=10.42.5.139:9995}: linkerd_app_core::errors: Failed to proxy request: error trying to connect: Connection refused (os error 111)
```
destination-proxy:
```text
[     0.007821s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.cluster.local. type: SRV class: IN
[     0.010546s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.cluster.local. type: SRV class: IN
[     2.010589s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.cluster.local. type: SRV class: IN
[     2.765354s]  INFO ThreadId(02) daemon:admin{listen.addr=0.0.0.0:4191}:accept{client.addr=127.0.0.1:58508 target.addr=127.0.0.1:4191}: linkerd_app_core::serve: Connection closed error=TLS detection timeout
[     4.014478s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.cluster.local. type: SRV class: IN
[     4.768364s]  INFO ThreadId(02) daemon:admin{listen.addr=0.0.0.0:4191}:accept{client.addr=127.0.0.1:58524 target.addr=127.0.0.1:4191}: linkerd_app_core::serve: Connection closed error=TLS detection timeout
[     6.009913s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.cluster.local. type: SRV class: IN
[     7.341143s]  WARN ThreadId(01) inbound:accept{client.addr=10.42.3.178:60680 target.addr=10.42.3.223:9996}: linkerd_app_core::errors: Failed to proxy request: error trying to connect: Connection refused (os error 111)
[     7.772956s]  INFO ThreadId(02) daemon:admin{listen.addr=0.0.0.0:4191}:accept{client.addr=127.0.0.1:58540 target.addr=127.0.0.1:4191}: linkerd_app_core::serve: Connection closed error=TLS detection timeout
```
identity
```text
Linkerd readiness check skipped: linkerd-await cannot block the identity controller
time="2021-03-19T21:25:18Z" level=info msg="running version stable-2.10.0"
time="2021-03-19T21:25:21Z" level=fatal msg="Failed to initialize identity service: Post \"https://10.43.0.1:443/apis/authorization.k8s.io/v1/selfsubjectaccessreviews\": read tcp 10.42.5.12:42080->10.43.0.1:443: read: connection reset by peer"
```
identity-proxy:
```text
62.289080s]  WARN ThreadId(01) outbound:accept{client.addr=10.42.5.12:33670 target.addr=10.210.151.100:6443}: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-dst-headless.linkerd.svc.cluster.local. type: SRV class: IN
[   762.323287s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=end of stream reached
[   762.431856s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=end of stream reached
```

services:
```text
linkerd-controller-api      ClusterIP   10.43.158.123   <none>        8085/TCP   20m
linkerd-dst                 ClusterIP   10.43.40.165    <none>        8086/TCP   20m
linkerd-dst-headless        ClusterIP   None            <none>        8086/TCP   20m
linkerd-identity            ClusterIP   10.43.33.182    <none>        8080/TCP   20m
linkerd-identity-headless   ClusterIP   None            <none>        8080/TCP   20m
linkerd-proxy-injector      ClusterIP   10.43.253.0     <none>        443/TCP    20m
linkerd-sp-validator        ClusterIP   10.43.181.209   <none>        443/TCP    20m
```

pods:
```text
linkerd-controller-7d7fb8f6b-fc9zq        0/2     CrashLoopBackOff   9          20m
linkerd-destination-6b5db68546-zl4p2      0/2     CrashLoopBackOff   7          16m
linkerd-identity-769d7b6c69-bdpjp         0/2     CrashLoopBackOff   7          15m
linkerd-proxy-injector-5686779df7-z6rmj   0/2     CrashLoopBackOff   9          20m
linkerd-sp-validator-7845d87fc4-rbdh6     0/2     CrashLoopBackOff   9          20m
```
(If the output is long, please create a [gist](https://gist.github.com/) and
paste the link here.)

#### `linkerd check` output

```text
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-cni-plugin
------------------
√ cni plugin ConfigMap exists
√ cni plugin PodSecurityPolicy exists
√ cni plugin ClusterRole exists
√ cni plugin ClusterRoleBinding exists
√ cni plugin Role exists
√ cni plugin RoleBinding exists
√ cni plugin ServiceAccount exists
√ cni plugin DaemonSet exists
√ cni plugin pod is running on all nodes

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust anchor

linkerd-webhooks-and-apisvc-tls
-------------------------------
√ proxy-injector webhook has valid cert
√ proxy-injector cert is valid for at least 60 days
√ sp-validator webhook has valid cert
√ sp-validator cert is valid for at least 60 days

linkerd-api
-----------
\ pod/linkerd-controller-7d7fb8f6b-fc9zq container public-api is not read
```

### Environment

- Kubernetes Version: v1.20.4
- Cluster Environment: Baremetal Rancher RKE (Cilium 1.9.5)
- Host OS: FlatCar 2765.2.1 
- Linkerd version: 2.10.1

**Comment by kleimkuhler** [comment]
@hwaastad I'm working on reproducing this, but have not had luck so far. I've been running a Cilium cluster through Kind as described in their [documentation](https://docs.cilium.io/en/v1.9/gettingstarted/kind/) and installing linkerd `stable-2.10.0`. Linkerd installs without any issue.

There is another issue #5679 where @networkop [commented](https://github.com/linkerd/linkerd2/issues/5679#issuecomment-808207561) about adding a port to `ignoreOutboundPorts`. It sounds like this "works", but I don't think that should be needed. I'd like to figure out why it was needed in that case, but it's easier to figure that out with a cluster demonstrating the issue.

For reference, my `describe` output of the `kuberenetes` service is:
```
Name:              kubernetes
Namespace:         default
Labels:            component=apiserver
                   provider=kubernetes
Annotations:       <none>
Selector:          <none>
Type:              ClusterIP
IP Families:       <none>
IP:                10.96.0.1
IPs:               <none>
Port:              https  443/TCP
TargetPort:        6443/TCP
Endpoints:         172.28.0.5:6443
Session Affinity:  None
Events:            <none>
```

I did do any extra configuration for my Linkerd installation. Do either of you have any more details about your installations that would help reproducing this? If you have the time, do you see this yourselves when installing on a Cilium cluster on Kind (steps for that in link above)?

**Comment by networkop** [comment]
@kleimkuhler it does look like a similar issue. Specifically, based on this error:
```
time="2021-03-19T21:25:21Z" level=fatal msg="Failed to initialize identity service: Post \"https://10.43.0.1:443/apis/authorization.k8s.io/v1/selfsubjectaccessreviews\": read tcp 10.42.5.12:42080->10.43.0.1:443: read: connection reset by peer"
```
I was initially baffled by this behaviour as well and asked a question on Cilium slack https://cilium.slack.com/archives/C1MATJ5U5/p1616758365247100

TLDR; cilium moves service load-balancing functionality inside the pod, so in your NAT redirect rules you have to account for any ports that endpoints may have.

**Comment by hwaastad** [comment]
@networkop, @kleimkuhler,  thanks for the info. And surly, when skipping outbound 6443 it's all ok.
I notice the 2.11 label so I'm looking forward to the fix.

**Comment by kleimkuhler** [comment]
Thanks for linking the slack conversation @networkop. Indeed after recreating a Kind cluster with Cilium and specifying ` --set hostServices.enabled=true` I see the same errors during installation.

Right now I'm not sure what the path forward is for Linkerd to work with this setting, but there are currently two solutions:
1. Add `6443` to the skip outbound port list
2. If you do not need host services enabled, then ` --set hostServices.enabled=false` will work

I'll swing back around to this issue when I understand a little more and maybe there is something that Linkerd can do so that its not affected by this setting.

**Comment by cortopy** [comment]
I've just installed linkerd to try it out for the first time and I'm experiencing the same issue. I use calico, not Cilium, so maybe not related to the CNI?

All the endpoints seem to exist:

```
NAME                        ENDPOINTS          AGE
linkerd-controller-api      10.42.6.103:8085   11m
linkerd-dst                 10.42.6.104:8086   11m
linkerd-dst-headless        10.42.6.104:8086   11m
linkerd-identity            10.42.1.27:8080    11m
linkerd-identity-headless   10.42.1.27:8080    11m
linkerd-proxy-injector      10.42.1.25:8443    11m
linkerd-sp-validator        10.42.1.26:8443    11m

```

but neither of the proxies seem to be able to resolve `linkerd-dst-headless` or `linkerd-identity-headless`

To make sure that dns resolution works, I've tested in a dnsutils container and this is what I get for dig

```
/ # dig linkerd-dst-headless.linkerd.svc.cluster.local

; <<>> DiG 9.11.6-P1 <<>> linkerd-dst-headless.linkerd.svc.cluster.local
;; global options: +cmd
;; Got answer:
;; WARNING: .local is reserved for Multicast DNS
;; You are currently testing what happens when an mDNS query is leaked to DNS
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 5430
;; flags: qr aa rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
; COOKIE: 563522a3bea8564e (echoed)
;; QUESTION SECTION:
;linkerd-dst-headless.linkerd.svc.cluster.local.        IN A

;; ANSWER SECTION:
linkerd-dst-headless.linkerd.svc.cluster.local. 5 IN A 10.42.6.104

;; Query time: 2 msec
;; SERVER: 10.43.0.10#53(10.43.0.10)
;; WHEN: Mon May 31 18:12:16 UTC 2021
;; MSG SIZE  rcvd: 149
```

**Comment by olix0r** [comment]
@cortopy What Kubernetes environment did you see this in? This should be a lot easier to track down if we figure out a way to reproduce it :)

**Comment by cortopy** [comment]
@olix0r of course! thanks for looking into this. It's on a cluster with a mix of bare metal and VM nodes. The cluster is managed with Rancher

A few more things:
- k8s version 1.20
- The nodes are connected via private network
- CNI is Calico
- Istio is already installed (I was testing for a possible migration) and working fine

**Comment by cortopy** [comment]
I gave another go at installing linkerd and see whether I could reproduce this issue and I have some logs that may be relevant.

This second time I was able to run linkerd containers. However, something very interesting happened last night. Traffic to all linkerd proxies started failing. My http services are responding with:

```
upstream connect error or disconnect/reset before headers. reset reason: connection failure
```

On inspection, the linkerd controller pod, had these logs:

```
time="2021-07-18T12:46:35Z" level=info msg="running version stable-2.10.2"
[     0.002589s]  INFO ThreadId(01) linkerd2_proxy::rt: Using single-threaded proxy runtime
[     0.003353s]  INFO ThreadId(01) linkerd2_proxy: Admin interface on 0.0.0.0:4191
[     0.003367s]  INFO ThreadId(01) linkerd2_proxy: Inbound interface on 0.0.0.0:4143
[     0.003370s]  INFO ThreadId(01) linkerd2_proxy: Outbound interface on 127.0.0.1:4140
[     0.003372s]  INFO ThreadId(01) linkerd2_proxy: Tap interface on 0.0.0.0:4190
[     0.003374s]  INFO ThreadId(01) linkerd2_proxy: Local identity is linkerd-controller.linkerd.serviceaccount.identity.linkerd.cluster.local
[     0.003377s]  INFO ThreadId(01) linkerd2_proxy: Identity verified via linkerd-identity-headless.linkerd.svc.cluster.local:8080 (linkerd-identity.linkerd.serviceaccount.identity.linkerd.cluster.local)
[     0.003379s]  INFO ThreadId(01) linkerd2_proxy: Destinations resolved via linkerd-dst-headless.linkerd.svc.cluster.local:8086 (linkerd-destination.linkerd.serviceaccount.identity.linkerd.cluster.local)
[     0.006871s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.cluster.local. type: SRV class: IN
[     0.011889s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.cluster.local. type: SRV class: IN
[     5.008581s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.cluster.local. type: SRV class: IN
[     7.095185s]  WARN ThreadId(01) server{port=9995}: linkerd_app_core::errors: Failed to proxy request: error trying to connect: Connection refused (os error 111) client.addr=144.76.87.53:55892
[    10.013706s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.cluster.local. type: SRV class: IN
[    11.878750s]  WARN ThreadId(01) server{port=9995}: rustls::session: Sending fatal alert AccessDenied    
[    11.984518s]  WARN ThreadId(01) server{port=9995}: rustls::session: Sending fatal alert AccessDenied    
[    12.187493s]  WARN ThreadId(01) server{port=9995}: rustls::session: Sending fatal alert AccessDenied    
[    12.606765s]  WARN ThreadId(01) server{port=9995}: rustls::session: Sending fatal alert AccessDenied    
[    13.108649s]  WARN ThreadId(01) server{port=9995}: rustls::session: Sending fatal alert AccessDenied    
[    13.614795s]  WARN ThreadId(01) server{port=9995}: rustls::session: Sending fatal alert AccessDenied    
[    14.122360s]  WARN ThreadId(01) server{port=9995}: rustls::session: Sending fatal alert AccessDenied    
[    14.546962s]  WARN ThreadId(01) server{port=4191}: rustls::session: Sending fatal alert AccessDenied    
[    14.623417s]  WARN ThreadId(01) server{port=9995}: rustls::session: Sending fatal alert AccessDenied    
[    14.649041s]  WARN ThreadId(01) server{port=4191}: rustls::session: Sending fatal alert AccessDenied    
[    14.874709s]  WARN ThreadId(01) server{port=4191}: rustls::session: Sending fatal alert AccessDenied    
[    15.010801s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.cluster.local. type: SRV class: IN
[    15.282387s]  WARN ThreadId(01) server{port=4191}: rustls::session: Sending fatal alert AccessDenied    
[    15.786476s]  WARN ThreadId(01) server{port=4191}: rustls::session: Sending fatal alert AccessDenied    
[    16.288156s]  WARN ThreadId(01) server{port=4191}: rustls::session: Sending fatal alert AccessDenied    
[    16.789526s]  WARN ThreadId(01) server{port=4191}: rustls::session: Sending fatal alert AccessDenied    
[    17.086943s]  WARN ThreadId(01) server{port=9995}: linkerd_app_core::errors: Failed to proxy request: error trying to connect: Connection refused (os error 111) client.addr=144.76.87.53:57756
[    17.290497s]  WARN ThreadId(01) server{port=4191}: rustls::session: Sending fatal alert AccessDenied    
[    17.510976s]  WARN ThreadId(01) server{port=9995}: linkerd_app_core::errors: Failed to proxy request: error trying to connect: Connection refused (os error 111) client.addr=144.76.87.53:57800
[    20.009340s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.cluster.local. type: SRV class: IN
[    21.877388s]  WARN ThreadId(01) server{port=9995}: rustls::session: Sending fatal alert AccessDenied    
[    21.979713s]  WARN ThreadId(01) server{port=9995}: rustls::session: Sending fatal alert AccessDenied    
[    22.190597s]  WARN ThreadId(01) server{port=9995}: rustls::session: Sending fatal alert AccessDenied    
[    22.607018s]  WARN ThreadId(01) server{port=9995}: rustls::session: Sending fatal alert AccessDenied    
[    23.108439s]  WARN ThreadId(01) server{port=9995}: rustls::session: Sending fatal alert AccessDenied    
[    23.609332s]  WARN ThreadId(01) server{port=9995}: rustls::session: Sending fatal alert AccessDenied    
[    24.110664s]  WARN ThreadId(01) server{port=9995}: rustls::session: Sending fatal alert AccessDenied    
[    24.547238s]  WARN ThreadId(01) server{port=4191}: rustls::session: Sending fatal alert AccessDenied    
[    24.613078s]  WARN ThreadId(01) server{port=9995}: rustls::session: Sending fatal alert AccessDenied    
[    24.656495s]  WARN ThreadId(01) server{port=4191}: rustls::session: Sending fatal alert AccessDenied    
[    24.863666s]  WARN ThreadId(01) server{port=4191}: rustls::session: Sending fatal alert AccessDenied    
[    25.008354s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.cluster.local. type: SRV class: IN
[    25.295471s]  WARN ThreadId(01) server{port=4191}: rustls::session: Sending fatal alert AccessDenied    
[    25.799949s]  WARN ThreadId(01) server{port=4191}: rustls::session: Sending fatal alert AccessDenied    
[    26.301917s]  WARN ThreadId(01) server{port=4191}: rustls::session: Sending fatal alert AccessDenied    
[    26.804076s]  WARN ThreadId(01) server{port=4191}: rustls::session: Sending fatal alert AccessDenied    
[    27.086511s]  WARN ThreadId(01) server{port=9995}: linkerd_app_core::errors: Failed to proxy request: error trying to connect: Connection refused (os error 111) client.addr=144.76.87.53:59306
[    27.306136s]  WARN ThreadId(01) server{port=4191}: rustls::session: Sending fatal alert AccessDenied    
[    27.507976s]  WARN ThreadId(01) server{port=9995}: linkerd_app_core::errors: Failed to proxy request: error trying to connect: Connection refused (os error 111) client.addr=144.76.87.53:59354
[    30.025476s]  INFO ThreadId(02) daemon:identity: linkerd_app: Certified identity: linkerd-controller.linkerd.serviceaccount.identity.linkerd.cluster.local
[ 82466.054309s]  WARN ThreadId(01) server{orig_dst=10.43.101.121:8086}: linkerd_reconnect::service: Service failed error=channel closed
[ 82466.054910s]  WARN ThreadId(01) server{orig_dst=10.43.101.121:8086}: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-dst-headless.linkerd.svc.cluster.local. type: SRV class: IN
[ 82466.063814s]  WARN ThreadId(01) server{orig_dst=10.43.101.121:8086}: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-dst-headless.linkerd.svc.cluster.local. type: SRV class: IN
[ 82471.056304s]  WARN ThreadId(01) server{orig_dst=10.43.101.121:8086}: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-dst-headless.linkerd.svc.cluster.local. type: SRV class: IN
[ 82476.057441s]  WARN ThreadId(01) server{orig_dst=10.43.101.121:8086}: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-dst-headless.linkerd.svc.cluster.local. type: SRV class: IN
```

For what I can see:

- The linkerd controller pod was restarted yesterday
- It took 30 seconds for the controller to be able to certify identity, and then after that it was all stable
- That's until nearly 23 hours later, when the pod was unable to resolve `linkerd-dst-headless.linkerd.svc.cluster.local`

Last night there was an issue with two nodes in the cluster. The issue is now fixed, but linkerd has entered into a bad state where any proxy that gets restarted is unable to resolve `linkerd-dst-headless.linkerd.svc.cluster.local`

Things that I have tried:
- Restarting all linkerd containers (this doesn't solve the issue)
- Updating the helm deployment (a new caBundle seems to be generated). This doesn't solve the issue either
- Weirdly, disabling cni in the helm chart and then re-enabling it solves the issue for the linkerd pods (i.e.: update helm chart twice). However, when I restart a pod with a http service I'm getting:

```
[     2.156934s]  INFO ThreadId(02) linkerd_app_core::serve: Connection closed error=TLS detection timed out
[     4.160928s]  INFO ThreadId(02) linkerd_app_core::serve: Connection closed error=TLS detection timed out
```

**Comment by cortopy** [comment]
Same thing happened today. It's very strange but whenever I have the DNS issues this solves it:

1. Update linkerd helm deployment with `cniEnabled` set as `false`
2. Wait for pods to become available
3. Update linkerd helm deployment with `cniEnabled` set as `true`

And now it all works as it should

**Comment by eric-aops** [comment]
Hi, 

The CNI workaround/implementation doesn't seem to solve the problem on our side. I'm still confronted with the DNS errors even though resolving works from a toolbox Pod in the same namespace. Only Linkerd Identity comes up, the other Pods just crash:
```
[   105.009907s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.apps-gke01.das-cld-stg.gcp.activeinfra.net. type: SRV class: IN
[   110.008892s]  WARN ThreadId(01) daemon:identity: linkerd_app_core::control: Failed to resolve control-plane component error=no record found for name: linkerd-identity-headless.linkerd.svc.apps-gke01.das-cld-stg.gcp.activeinfra.net. type: SRV class: IN
[   112.505406s]  WARN ThreadId(01) server{port=9995}: linkerd_app_core::errors: Failed to proxy request: error trying to connect: Connection refused (os error 111) client.addr=10.93.176.13:52588
Using a toolbox container I verified that the domain name is resolvable from within the linkerd namespace:
root@toolbox:/# dig linkerd-identity-headless.linkerd.svc.apps-gke01.das-cld-stg.gcp.activeinfra.net

; <<>> DiG 9.11.5-P4-5.1ubuntu2-Ubuntu <<>> linkerd-identity-headless.linkerd.svc.apps-gke01.das-cld-stg.gcp.activeinfra.net
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 22955
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;linkerd-identity-headless.linkerd.svc.apps-gke01.das-cld-stg.gcp.activeinfra.net. IN A

;; ANSWER SECTION:
linkerd-identity-headless.linkerd.svc.apps-gke01.das-cld-stg.gcp.activeinfra.net. 30 IN	A 10.210.2.44

;; Query time: 2 msec
;; SERVER: 169.254.20.10#53(169.254.20.10)
;; WHEN: Thu Sep 09 00:28:55 UTC 2021
;; MSG SIZE  rcvd: 205

root@toolbox:/
```
Tried with FW updated to open all ports from controlplane and between nodes but no joy.

Any pointers or assistance is greatly appreciated. 

Kind regards, 

Eric V.

**Comment by sysnasri** [comment]
I have had a similar issue, workaround I have to changed  cluster.local to the name of our kubernetes cluster which is cluster.something

**Comment by olix0r** [comment]
> I have had a similar issue, workaround I have to changed cluster.local to the name of our kubernetes cluster which is cluster.something

Yes, you need to configure linkerd to know about custom cluster domains when they are used. Though, it appears that this issue also describes other situations where this behavior can arise.

Unfortunately we haven't had enough information to reproduce or diagnose these other scenarios.

**Comment by adleong** [comment]
hi @KuasarCloud, would you mind please filing a new issue with detailed reproduction steps including specifically the steps to set up the cluster using kubeadm and cilium?  If we can reproduce this issue, we will be much more effective in being able to fix it.
